girder/wsgi.py

```diff
@@ -18,14 +18,24 @@
 ###############################################################################
 
 import cherrypy
-from girder.utility import config, server
+import girder
+import logging.handlers
+from girder.utility import server
 
-config.loadConfig()  # Read the config files first, so we can override some values
 cherrypy.config.update({'engine.autoreload.on': False,
                         'environment': 'embedded'})
 cherrypy.config['server'].update({'cherrypy_server': False,
                                   'disable_event_daemon': True})
 
+# TODO The below line can be removed if we do away with girder.logprint
+girder._quiet = True  # This means we won't duplicate messages to stdout/stderr
+_formatter = girder.LogFormatter('[%(asctime)s] %(levelname)s: %(message)s')
+_handler = logging.handlers.StreamHandler()  # Send logs to stderr
```

Comment:
Per the (slightly bitter) descriptions [here](https://modwsgi.readthedocs.io/en/develop/user-guides/application-issues.html#writing-to-standard-output) and [here](https://modwsgi.readthedocs.io/en/develop/user-guides/debugging-techniques.html#apache-error-log-files) it looks like we should be writing to the WSGI environ value `'wsgi.errors'`, which is a stream. However, while CherryPy does provide this via `request.wsgi_environ['wsgi.errors']`, the value is going to reset on every request, it doesn't allow us to log messages that occur outside of a request (like server startup), and it's probably going to be awkward within Girder to attach a new handler at the start of every request and detach one once it's over (though it's probably possible if we dig into CherryPy's wildly-undocumented [WSGI middleware capabilities](https://github.com/cherrypy/cherrypy/blob/master/cherrypy/_cpwsgi.py).

Accordingly, I think the second best we can do is to log to `sys.stderr` here. It looks like we should attempt to never write to `sys.stdout` if we want Apache's mod_wsgi to be totally happy (and that's what's used by AWS Beanstalk by default).

Reply:
I can't tell if this comment is requesting a change here or affirming that I made the correct decision :)

Reply:
We should change it to log to `sys.stderr` only.

I honestly don't understand how to interpret this [statement from the docs](https://docs.python.org/2/library/logging.handlers.html#streamhandler) means:
> If *stream* is specified, the instance will use it for logging output; otherwise, *sys.stderr* will be used.

but I think we should just have:
```
_handler = logging.handlers.StreamHandler(sys.stderr)
```
instead.

Reply:
It means `sys.stderr` is the default if none is passed, so your proposed change won't affect behavior, but it's worthwhile for readability.

Reply:
Oops, somehow I totally misread those docs, and thought it was talking about `sys.stdout` also in some way.

Reply:
However, I've dug into CherryPy a bit more, and they do provide a [`WSGIErrorHandler`](https://github.com/cherrypy/cherrypy/blob/master/cherrypy/_cplogging.py#L441), which we could instantiate and attach ourselves to the base Girder logger (CherryPy also has an API to attach one to the CherryPy loggers). However, as mentioned, this would still not capture log output from times outside of a request context.

What things does Girder log outside of a request? Obviously plugin enablement, but is there anything else? Additionally, how much do we care about Apache being able to associate Girder's logs with the particular virtual host that it's running under? Do we anticipate running Girder with multiple virtual host web apps on the same Apache server?

Reply:
> What things does Girder log outside of a request? Obviously plugin enablement, but is there anything else? 

This is kind of a design question -- are the logs meant to be usable in all contexts of Girder, or only when it's being used as a web server? I could go either way on this.

> Additionally, how much do we care about Apache being able to associate Girder's logs with the particular virtual host that it's running under? Do we anticipate running Girder with multiple virtual host web apps on the same Apache server?

I'm not sure, but it would be good to support this in the most first-class way possible.

Reply:
@brianhelba @girder/developers thoughts on this? Should I attach this handler instead of the StreamHandler?